### ISSUE TITLE
services.influxdb service requires java ?

### ISSUE THREAD

**Issue by netixx** [NixOS/nixpkgs#59124]
## Issue description
Using `services.influxdb.enable = true` causes `jdk` to be installed, requiring options for java licence and unfree packages.

### Steps to reproduce

Configure influxdb in configuration.nix

```nix
services.influxdb.enable = true
```

## Technical details
```
# nixos-rebuild switch          
building Nix...
building the system configuration...
error: You must accept the Oracle Binary Code License Agreement for Java SE at
https://www.oracle.com/technetwork/java/javase/terms/license/index.html
by setting nixpkgs config option 'oraclejdk.accept_license = true;'

(use '--show-trace' to show detailed location information)
```
 - system: `"aarch64-linux"`
 - host os: `Linux 4.19.34, NixOS, 18.09.2478.f9576d0cce2 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(root): `"nixos-18.09.2478.f9576d0cce2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by netixx** [comment]
`influxdb` is a go package, and issuing `nix-shell -p influxdb` does not yield this behaviour, and the issuing binaries works.

**Comment by erictapen** [comment]
Let's build a basic system with that service.

```nix
(import <nixpkgs/nixos/lib/eval-config.nix> {
  system = "x86_64-linux";
  modules = [
    ({config, pkgs, ...}:{
      services.influxdb.enable = true;

      # stuff we need to match some assertions.
      fileSystems."/" =
        { device = "/dev/disk/by-uuid/bla";
          fsType = "ext4";
        };
      boot.loader.grub.device = "/dev/sda";

    })
  ];
}).config.system.build.toplevel
```

Answering the question "Why depends this on that at build time?" isn't too easy to answer with the current Nix tooling, but I provided a workaround in https://github.com/NixOS/nix/issues/1918#issuecomment-444110195. So lets do

```bash
$ nix-shell -p graphviz
$ nix-store -q --graph $(nix-instantiate default.nix) \ # default.nix is that system definition above
  | dijkstra -da $(nix-instantiate '<nixpkgs>' -A openjdk) \
  | gvpr -c 'N[dist>1000.0]{delete(NULL, $)}' \
  | dot -Tpng > graph.png
```

On `nixos-18.09` this results in

![graph](https://user-images.githubusercontent.com/11532355/55717655-11fcff80-59fa-11e9-88fb-00edb5f14bd6.png)

So it turns out, that the systemd service definition `unit-influxdb.service` is what pulls in `collectd` and with that `openjdk`. Thats why you didn't see `openjdk` when doing `nix-shell`. [This](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/databases/influxdb.nix#L71) might be the line at fault here.

**Comment by netixx** [comment]
Right, thank you for this analysis! This is one of the frustrating things with nixos, because it is not easy to know why some dependency is installed. Those sorts of commands should be in the (advanced) manual, or even be a `nixos-*` command.

The one think I am missing is why in my case is the proprietary jdk installed ? Is openjdk not available for aarch64 (on nix) ?

**Comment by erictapen** [comment]
I missed that you are on `aarch64-linux` and that the package in question is `oraclejdk`. I updated my answer accordingly.

**Comment by arianvp** [comment]
Nix 2.0 has a high level `nix why-depends` command that should in theory serve this usecase

**Comment by erictapen** [comment]
@arianvp: Unfortunately `wyh-depends` only supports run-time dependencies, so the link in question wouldn't be observable with this command. See also the linked issue https://github.com/NixOS/nix/issues/1918 for context. I'd love to see that functionality in `why-depends`.

**Comment by netixx** [comment]
@zimbatm @offlinehacker can you think of way to not pull `collectd` (and java) when using `influxdb` ?

I was thinking of modifying this block to contionnaly define the `typesdb` attribute based on a config switch, but I am not sure this is the right way to do this...
```nix
collectd = [{
      enabled = false;
      typesdb = "${pkgs.collectd-data}/share/collectd/types.db";
      database = "collectd_db";
      bind-address = ":25826";
    }];
```

**Comment by netixx** [comment]
At the moment I am using the following snippet to remove the option (`updateRecursive` is used behind the wheel and `collectd` is a list, which is why it works):   `config.services.influxdb.extraConfig.collectd = [{enable=false;}];`

I find this config strange since enabling collectd using `extraConfig.collectd = [{enable=true}};` will remove typesdb option and associated config...

**Comment by arianvp** [comment]
Still an issue

**Comment by zimbatm** [comment]
Is it even necessary to set a value? Couldn't you have `collectd = []` instead?
